Allow configs for reporters; Rubocop can be passed additional args

jonknapp opened the pull request:
This allows solargraph reporters to have configurations passed to them from a `.solargraph` config file. They are passed a hash of the options, which should all be empty hashes except for Rubocop which can now take an `arguments` param for additional parameters to add to commands ran.

It allows support for https://github.com/castwide/solargraph/issues/93.

For example:

```
---
reporters:
- require_not_found
- rubocop:
    arguments:
    - --force-exclusion
```

would run Rubocop with the [`--force-exclusion` option](https://docs.rubocop.org/en/latest/basic_usage/#other-useful-command-line-flags). The `arguments` portion of the reporter config should be an array of strings.

castwide commented:
Thanks for the quick work. Looks solid to me at first glance. My only concern is that passing arguments as a single string could lead to mangled escape sequences. That's the main reason for passing them as an array.

jonknapp commented:
It's not a problem to change `arguments` to be an array if you feel that will work better. Feel free to comment on anything else if you get some time to look it over again.

I'll add info to the README and update the code to handle an array of arguments instead of a string.

jonknapp commented:
I added some documentation and changed the `arguments` part of the reporter config to be an array of strings instead of a single string that we parse. If anything else is needed, let me know. Otherwise, I'm going to look at the VSCode plugin and see what needs to be done there to support this new feature.

castwide commented:
Thanks. I merged the PR into a new branch `rubocop-custom-arguments`, since there's a patch update I expect to release in the next few days. The next release will be 0.28.3, and I expect this PR to be released in 0.29.0.

The update to VSCode will probably require a change to pass client configuration data to the reporters, most likely in `LanguageServer::Host#diagnose`.

PiotrKolaczRM commented:
Will this be merged into `master`?
